### ISSUE TITLE
Why doing Key and Query Masking ?

### ISSUE THREAD

**Issue by odie2630463** [Kyubyong/transformer#3]
Nice Job!
But in paper doesn't describe Key and Query Masking , can you give me some hint about that ?
thanks!

**Comment by Kyubyong** [comment]
The encoder leaves artifacts, which are non-zeros, for paddings. It doesn't make sense, a query attends to those, so before applying softmax for getting score, I overwrote them with very very small numbers. As a results they will have score 0. Likewise, queries for paddings should not have any values, so they are masked with zeros.

**Comment by zhedongzheng** [comment]
But why do you put (-2**32+1), it is a huge num, not small nums
thanks for your time

**Comment by darongliu** [comment]
The usage of the masking is judged by whether the summation of the keys or queries at the last dimension is zero. However, the padding part, which is originally embedded with zeros, is added by positional embedding, so they will never be zero. Thanks!!!!

**Comment by zhedongzheng** [comment]
@darongliu You have raised a very important point. We have to create a mask before the positional encoding, otherwise the whole padding doesn't work.

**Comment by sunnnnnnnny** [comment]
@zhedongzheng  Does this mean that this piece has a problem?

**Comment by darongliu** [comment]
No, I think it will not cause great performance drop. It is still a good implementation.

**Comment by sunnnnnnnny** [comment]
@darongliu  Doesn't that mean that the query mask and key mask are useless?

**Comment by Life-0-1** [comment]
Query masking is unnecessary? cause the padded query will be masked out by next block's key masking?

**Comment by achillesliu** [comment]
> @darongliu Doesn't that mean that the query mask and key mask are useless?

It's not working. Check https://github.com/Kyubyong/transformer/issues/33
